test/framework/image_preloading.go

```diff
@@ -0,0 +1,220 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/cluster-api/test/framework"
+	. "sigs.k8s.io/cluster-api/test/framework/ginkgoextensions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func LoadImagesFunc(ctx context.Context) func(clusterProxy framework.ClusterProxy) {
+	sourceFile := os.Getenv("DOCKER_IMAGE_TAR")
+	Expect(sourceFile).ToNot(BeEmpty(), "DOCKER_IMAGE_TAR must be set")
+
+	loader := imagePreloader{
+		sourceFile: sourceFile,
+	}
+
+	return func(clusterProxy framework.ClusterProxy) {
+		loader.ImagesToCluster(ctx, clusterProxy)
+	}
+}
+
+type imagePreloader struct {
+	sourceFile string
+}
+
+// ImagesToCluster deploys a privileged daemonset and uses it to stream-load container images.
+func (loader *imagePreloader) ImagesToCluster(ctx context.Context, clusterProxy framework.ClusterProxy) {
+	daemon, daemonMutateFn, daemonLabels := getPreloadDaemonset()
+	ctrlClient := clusterProxy.GetClient()
+
+	// Create Daemonset
+	_, err := controllerutil.CreateOrPatch(ctx, ctrlClient, daemon, daemonMutateFn)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Wait for DaemonSet to be available.
+	waitForDaemonSetAvailable(ctx, waitForDaemonSetAvailableInput{Getter: ctrlClient, Daemonset: daemon}, time.Minute*3, time.Second*10)
+
+	// List all pods and load images via each found pod.
+	pods := &corev1.PodList{}
+	Expect(ctrlClient.List(
+		ctx,
+		pods,
+		client.InNamespace(daemon.Namespace),
+		client.MatchingLabels(daemonLabels),
+	)).To(Succeed())
+
+	errs := []error{}
+	for j := range pods.Items {
+		pod := pods.Items[j]
+		Byf("Loading images to node %s via pod %s", pod.Spec.NodeName, klog.KObj(&pod))
+		if err := loader.imagesViaPod(ctx, clusterProxy, pod.Namespace, pod.Name, pod.Spec.Containers[0].Name); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	Expect(kerrors.NewAggregate(errs)).ToNot(HaveOccurred())
+}
+
+func (loader *imagePreloader) imagesViaPod(ctx context.Context, clusterProxy framework.ClusterProxy, namespace, podName, containerName string) error {
+	// Open source tar file.
+	reader, writer := io.Pipe()
+	file, err := os.Open(loader.sourceFile)
+	if err != nil {
+		return err
+	}
+
+	// Use go routine to pipe source file content into then stdin.
+	go func(file *os.File, writer io.WriteCloser) {
+		defer writer.Close()
+		defer file.Close()
+		_, _ = io.Copy(writer, file)
```

Comment:
Should we handle the err? Otherwise how do we know it worked?

Reply:
We would know depending on the below exec was successful or not.

Reply:
@chrischdi WDYT about logging the error at least? Just might make it easier to debug if it fails